contracts/Staking.sol

```diff
@@ -0,0 +1,559 @@
+pragma solidity 0.6.4;
+
+import "./System.sol";
+import "./lib/Memory.sol";
+import "./lib/BytesToTypes.sol";
+import "./interface/IApplication.sol";
+import "./interface/ICrossChain.sol";
+import "./interface/IParamSubscriber.sol";
+import "./interface/ITokenHub.sol";
+import "./lib/CmnPkg.sol";
+import "./lib/SafeMath.sol";
+import "./lib/RLPEncode.sol";
+import "./lib/RLPDecode.sol";
+import "./interface/IStaking.sol";
+
+contract Staking is IStaking, System, IParamSubscriber, IApplication {
+  using SafeMath for uint256;
+  using RLPEncode for *;
+  using RLPDecode for *;
+
+  // Cross-Chain Stake Event type
+  uint8 public constant EVENT_DELEGATE = 0x01;
+  uint8 public constant EVENT_UNDELEGATE = 0x02;
+  uint8 public constant EVENT_REDELEGATE = 0x03;
+  uint8 public constant EVENT_DISTRIBUTE_REWARD = 0x04;
+  uint8 public constant EVENT_DISTRIBUTE_UNDELEGATED = 0x05;
+
+  uint32 public constant ERROR_UNKNOWN_PACKAGE_TYPE = 101;
+  uint32 public constant ERROR_WITHDRAW_BNB = 102;
+
+  uint256 constant public TEN_DECIMALS = 1e10;
+
+  uint256 public constant INIT_ORACLE_RELAYER_FEE = 2e16; //TODO
+  uint256 public constant INIT_MIN_DELEGATION = 100 * 1e18;
+
+  uint256 public oracleRelayerFee;
+  uint256 public minDelegation;
+
+  mapping(address => uint256) delegated; // delegator => totalAmount
+  mapping(address => mapping(address => uint256)) delegatedOfValidator; // delegator => validator => amount
+  mapping(address => uint256) distributedReward; // delegator => reward
+  mapping(address => mapping(address => uint256)) pendingUndelegateTime; // delegator => validator => minTime
+  mapping(address => uint256) undelegated; // delegator => totalUndelegated
+  mapping(address => mapping(address => mapping(address => uint256))) pendingRedelegateTime; // delegator => srcValidator => dstValidator => minTime
+
+  bool internal locked;
+
+  modifier noReentrant() {
+    require(!locked, "No re-entrancy");
+    locked = true;
+    _;
+    locked = false;
+  }
+
+  modifier tenDecimalPrecision(uint256 amount) {
+    require(msg.value%TEN_DECIMALS==0 && amount%TEN_DECIMALS==0, "precision loss in conversion");
+    _;
+  }
+
+  modifier initParams() {
+    if (!alreadyInit) {
+      oracleRelayerFee = INIT_ORACLE_RELAYER_FEE;
+      minDelegation = INIT_MIN_DELEGATION;
+      alreadyInit = true;
```

Comment:
Can we define `alreadyInit` var in this file? so that when we need add new variable , we get high confidence that the storage slot won't break.

Reply:
Done